### PR TITLE
add kv_cache to LLM

### PR DESCRIPTION
**Pull request by vchiley**
This pr include past_key_values (ie kv_cache) in the LLM so that inference can be accelerated.
We also become explicit about how we apply `padding_mask` for querys and keys.

Shoutout: @dakinggg for working through some of this with me.

cc @dskhudia @alextrott16 @samhavens for after training / inference

### CODE REVIEW

**Comment by honglu2875** [comment]
@vchiley Cached kv values shift the positions as well. Maybe you want to shift the position embeddings as well?
https://github.com/mosaicml/examples/blob/e98e6750ee8d917b60e1090845aba80fab816c2d/examples/llm/src/models/mosaic_gpt.py#L148
Compare with this in HF
https://github.com/huggingface/transformers/blob/60d51ef5123d949fd8c59cd4d3254e711541d278/src/transformers/models/gpt2/modeling_gpt2.py#L801

In our fork, the relevant part looks like the following:
```python
        if past_key_values is None:
            past_key_values = [None] * self.cfg.n_layers
            past_position = 0
        else:
            assert len(past_key_values) == self.cfg.n_layers
            # get the key tensor whose spec should be (batch, seq, n_head, head_dim), and
            # collect the `seq`, so that we shift the position embedding later.
            past_position = past_key_values[0][0].size(1)

        tok_emb = self.transformer.wte(input_ids)  # type: ignore
        if self.alibi:
            x = tok_emb
        else:
            if S + past_position > self.cfg.max_seq_len:
                raise ValueError(
                    f'Cannot forward input with past sequence length {past_position} and current sequence length '
                    f'{S + 1}, this model only supports total sequence length <= {self.cfg.max_seq_len}.'
                )
            pos = torch.arange(past_position, S + past_position, dtype=torch.long,
                               device=input_ids.device).unsqueeze(0)
            pos_emb = self.transformer.wpe(pos)  # type: ignore
            x = tok_emb + pos_emb
```

**Review by dakinggg — APPROVED** [review]
LGTM, can you train a model and make sure nothing is broken?

**Comment by dskhudia** [comment]
Could you explain the reason for separating out query_padding_mask?

File: examples/llm/src/models/layers/attention.py [hunk 1]
```diff
@@ -266,12 +302,23 @@ def forward(self,
...
+            query_padding_mask = key_padding_mask[:, -query.size(1):]
+
         if self.attn_qk_ln:
             # Applying layernorm to qk
             dtype = query.dtype
             query = self.q_ln(query).to(dtype)
             key = self.k_ln(key).to(dtype)
 
+        if past_key_value is not None:
+            if past_key_value:
```
[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:316]: if you're generating the first element, past_key_value will be an empty tuple.
past_key_value will not be None, but it will eval to false.

[the remark this answers is not in the document]
Comment (dakinggg) [reply examples/llm/src/models/layers/attention.py:316]: could you check explicitly instead of falseyness? e.g. `len(past_key_value) == 0`. just for clarity and to avoid silent bugs.

**Comment by vchiley** [comment]
@dskhudia this formulates a generic attn fn
the queries are potentially not the same as the keys/value and will need their own padding_mask.
This is useful for left padded inputs since past tokens influence future tokens, we NEED to mask them out.

File: examples/llm/src/models/layers/attention.py [hunk 2]
```diff
@@ -179,16 +196,31 @@ def triton_flash_attn_fn(
...
+        if query_padding_mask is not None:
+            s_q = query_padding_mask.size(1)
+        if key_padding_mask is not None:
+            s_k = key_padding_mask.size(1)
 
         if attn_bias is not None:
             attn_bias = attn_bias.expand(b_size, -1, -1, -1)
         else:
-            attn_bias = query.new_zeros(b_size, 1, 1, s_k)
+            attn_bias = query.new_zeros(b_size, 1, s_q, s_k)
```
Comment (dskhudia) [inline examples/llm/src/models/layers/attention.py:215]: aren't s_q and s_k undefined if query_padding_mask and key_padding_mask are not None

File: examples/llm/src/models/layers/attention.py [hunk 3]
```diff
@@ -256,6 +291,7 @@ def __init__(self, cfg: DictConfig, device: Optional[str] = None):
 
     def forward(self,
                 x,
+                past_key_value=None,
```
Comment (dskhudia) [inline examples/llm/src/models/layers/attention.py:294]: Keep key_padding_mask as the second argument? As that's the most common one during training.

File: examples/llm/src/models/mosaic_gpt.py [hunk 4]
```diff
@@ -143,7 +145,21 @@ def forward(self,
         if self.alibi:
             x = tok_emb
         else:
-            pos = torch.arange(0, S, dtype=torch.long,
+            past_position = 0
+            if past_key_values is not None:
+                assert len(past_key_values) == self.cfg.n_layers
```
[the remark this answers is not in the document]
Comment (dskhudia) [reply examples/llm/src/models/mosaic_gpt.py]: Assert is better from export point of view (supported by torchscript).

**Review by dskhudia — APPROVED** [review]

File: examples/llm/src/models/mosaic_gpt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply examples/llm/src/models/mosaic_gpt.py]: ah ok, totally fine to leave assert if needed for torchscript

File: examples/llm/src/models/layers/attention.py (hunk 2 above)
Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:215]: [They default to 1](https://github.com/mosaicml/examples/pull/244/files/0d459af1ac60055092e1e0c972a9e29d5b26fe55#diff-22824d61d86c454e9e92d5ea2b1a233d8d972160e01d2261f3dd8f18c307c4f5R206)
and everything works

File: examples/llm/src/models/mosaic_gpt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/src/models/mosaic_gpt.py]: I already changed it.
Should we change all raise back to assert?
Doesn't our style guide ask for raise not assert?

File: examples/llm/src/models/layers/attention.py (hunk 3 above)
Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:294]: x and past_key_value are activations so I kept them together.
key_padding_mask is a mask that defines what we do with the activations so I moved that after the activations.

Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:294]: its also how [hf does it](https://github.com/huggingface/transformers/blob/c07a02a4b7892edfee22cbe57d3cdd9e10ae7a4d/src/transformers/models/gpt2/modeling_gpt2.py#L1050C11-L1065).
keeping past_key_value as second arg.

File: examples/llm/src/models/mosaic_gpt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/src/models/mosaic_gpt.py]: Either way, deciding to change our style guide seems outside the scope of this PR. Keeping raise for now.

[the remark this answers is not in the document]
Comment (vchiley) [reply examples/llm/src/models/mosaic_gpt.py]: Also @dakinggg noted [asserts can be turned off](https://stackoverflow.com/questions/1273211/disable-assertions-in-python) 🤯

**Comment by vchiley** [comment]
Note: we should have a conversation about if all `raise Error` should be changed to `assert` (see [here](https://github.com/mosaicml/examples/pull/244#discussion_r1142772540))

File: examples/llm/src/models/mosaic_gpt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dskhudia) [reply examples/llm/src/models/mosaic_gpt.py]: We can change it later if we run into issues with export. torchscript only supports a subset of python and I am not sure if raise is part of that.

File: examples/llm/src/models/layers/attention.py (hunk 3 above)
Comment (dskhudia) [reply examples/llm/src/models/layers/attention.py:294]: They don't have key_padding_mask though. Anyway just a suggestion so what you have is fine.

File: examples/llm/src/models/layers/attention.py (hunk 2 above)
Comment (dskhudia) [reply examples/llm/src/models/layers/attention.py:215]: They won't default to one when both query_padding_mask and key_padding_mask are given and will be undefined. Do we have a test for this case?

File: examples/llm/src/models/mosaic_gpt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (samhavens) [reply examples/llm/src/models/mosaic_gpt.py]: @dskhudia `raise` is supported, but `try/except` isn't, https://pytorch.org/docs/stable/jit_language_reference_v2.html#the-raise-statement

File: examples/llm/src/models/layers/attention.py (hunk 2 above)
Comment (vchiley) [reply examples/llm/src/models/layers/attention.py:215]: are s_k or s_q used outside of the `if query_padding_mask is not None or key_padding_mask is not None:` context?

Comment (dskhudia) [reply examples/llm/src/models/layers/attention.py:215]: nvm. I misread line 215 as outside the if context.
